### ISSUE TITLE
[CURATOR-627] Curator remains in SUSPENDED state ignoring session timeout or network recovery

### ISSUE THREAD

**Issue by jira-importer** [apache/curator#1144]
<h1><a name="Description"></a>Description</h1>

<p>Our program encountered a problem where curator got stuck in the SUSPENDED state. It kept injecting "session expire" event, generating logs like this every 15s (same as session timeout):</p>
<blockquote><p>Session timeout has elapsed while SUSPENDED. Injecting a session expiration</p></blockquote>
<p>However, curator never transited to the LOST state after session timeout, nor transied to the RECONNECTED state upon network recovery. </p>

<p> </p>

<p>According to logs containing "zookeeper" or "curator" (see attachment "log4j.log"), related events are as follows:</p>

<p>(All events happened in date 2021.12.05)</p>
<ul>
	<li>23:34:07,122: ZooKeeper not heard from server, thus curator transited to the SUSPENDED state.</li>
</ul>

<ul>
	<li>23:34:09,128: ZooKeeper opening socket to zk server.</li>
</ul>

<ul>
	<li>23:34:22,223: curator injected a "session expiration" event, but had not transited to the LOST state at that time; never transited in the future.</li>
</ul>

<ul>
	<li>23:34:22,885: ZooKeeper opened socket to zk server, and received reponse indicating session expiration; thus, the current ZooKeeper object was closed. However, another ZooKeeper object was not created as expected.</li>
</ul>

<ul>
	<li>23:34:37,224 and later: curator kept injecting "session expiration" events every session timeout (15s). The log was filled with lines like this: "Session timeout has elapsed while SUSPENDED. Injecting a session expiration."</li>
</ul>

<p>In summary, curator stayed in the SUSPENED state and never transited to the LOST or RECONNECTED state. Besides, the underlying ZooKeeper object was never recreated according to logs and the "jstack" command. For a more detailed analysis, please refer to attachment "log analysis.md".</p>

<p> </p>

<p>The problem above was encountered only once in our testing environment within months, and has never occurred in our production enviromnent. We failed to find out how to reproduce, but suspect there is a racing condition when these events happen simultaneously: while curator is injecting a "session expire" event, the underlying "ClientCnxnSocketNIO" has just reconnected to a zk server and got a "session expire" response.</p>
<h1><a name="Environment"></a>Environment</h1>

<p>OS: Linux version 4.17.3-1.el7.elrepo.x86_64</p>

<p> </p>

<p>Project type: maven</p>

<p>Curator version: 4.2.0</p>

<p> </p>

<p>Session timeout: 15s, which is equal to that negotiated with the server.</p>

<p>Retry policy: no retry, where RetryPolicy#allowRetry() always returns false. That's because we handle retry at application level. Write operations should NOT be retried immediately after reconnected, until some extra validations pass.</p>

<p> </p>

<p>Our program creates exactly one "CuratorFramework" instance connecting to the only zk server in the testing environment. The server's connection string is "6x.xx.xx.27:2181", with some digits consealed for security. Ip addresses in logs are also such processed. </p>

<p> </p>

<p>ZkClient is imported for forward compatiblility with some legacy codes (still dependent on ZkClient), but it is bridged to curator instead of connecting to zk server itself. For more details about dependencies, please refer to file "pom.xml" in attachments.</p>
<h1><a name="Attachments"></a>Attachments</h1>

<p>log4j.log: log of our program, with ip addresses and zk paths consealed.</p>

<p>log analysis.md: a more detailed analysis about the logs.</p>

<p>pom.xml: part of file "pom.xml", showing zookeeper-related dependencies.</p>

---
<details><summary><i>Originally reported by <a title="Bai Yu" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=Bai Yu">Bai Yu</a>, imported from: <a href="https://issues.apache.org/jira/browse/CURATOR-627" target="_blank">Curator remains in SUSPENDED state ignoring session timeout or network recovery</a></i></summary>
<i><ul>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2023-04-02T07:48:57+01:00
<li><b>imported</b>: 2025-01-21
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=thenatog">thenatog</a>:</i>
<p>Some of our users have also seen this in Apache NiFi which uses Curator 4.2.0 (we're scheduled to upgrade to 5.2.0 in our next release):</p>

<p><a href="https://issues.apache.org/jira/browse/NIFI-9559" class="external-link" rel="nofollow">https://issues.apache.org/jira/browse/NIFI-9559</a></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=absolutesantaja">absolutesantaja</a>:</i>
<p>I think this is a duplicate of <a href="https://issues.apache.org/jira/browse/CURATOR-561" title="Race condition preventing reconnection" class="issue-link" data-issue-key="CURATOR-561"><del>CURATOR-561</del></a> or at least related. It appears that in 5.2.0 once suspended nothing ever actually forces it to reconnect.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=msohn">msohn</a>:</i>
<p>It seems we are hit by this bug as well.</p>

<p>We use a curator cache to keep track of the cluster state in zookeeper of an ensemble of Gerrit servers.</p>

<p>We tested a network partition where one of the application nodes can't connect to zookeeper anymore.<br/>
When the network partition ended curator seems not to reconnect and hangs in state "CONNECTING" with isConnected is false.</p>

<p>Is there an ETA for release 5.2.1 ?</p>
